src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,189 @@
+        private Action<string> _writeEvent;
+
+        public ConsoleWriter(IEnvironment environment, Action<Exception> unhandledExceptionHandler)
+            : this(environment, unhandledExceptionHandler, consoleBufferTimeout: DefaultConsoleBufferTimeout, autoStart: true)
+        {
+        }
+
+        internal ConsoleWriter(IEnvironment environment, Action<Exception> exceptionHandler, TimeSpan consoleBufferTimeout, bool autoStart)
+        {
+            bool consoleEnabled = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingDisabled) != "1";
+
+            if (consoleEnabled)
+            {
+                // We are going to used stdout, but do we write directly or use a buffer?
+                _consoleBuffer = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferSize) switch
+                {
+                    "-1" => Channel.CreateUnbounded<string>(),              // buffer size of -1 indicates that buffer should be enabled but unbounded
+                    "0" => null,                                            // buffer size of 0 indicates that buffer should be disabled
+                    var s when int.TryParse(s, out int i) && i > 0 => Channel.CreateBounded<string>(i),
+                    _ => Channel.CreateBounded<string>(DefaultBufferSize),  // default behavior is to use buffer with default size
+                };
+
+                if (_consoleBuffer == null)
+                {
+                    _writeEvent = Console.WriteLine;
+                }
+                else
+                {
+                    _writeEvent = WriteToConsoleBuffer;
+                    _consoleBufferTimeout = consoleBufferTimeout;
+                    if (autoStart)
+                    {
+                        bool batched = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferBatched) switch
+                        {
+                            "0" => false,     // disable batching by setting to 0
+                            _ => true,        // default behavior is batched
+                        };
+
+                        StartProcessingBuffer(batched);
+                    }
+                }
+            }
+            else
+            {
+                _writeEvent = (string s) => { };
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public void WriteHandler(string evt)
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
+                {
+                    // Buffer is currently full, wait until writing is permitted.
+                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
```

Comment:
This cast is not ideal and may cause an extra allocation in the `IValueTaskSource` case, which would be counterproductive for the performance gains. However, not sure how to address this. But I do suggest the following:

``` CSharp
try
{
    if (!_consoleBuffer.Writer.TryWrite(ext))
    {
        ValueTask write = _consoleBuffer.Writer.WriteAsync(ext);
        if (write.IsCompleted)
        {
            write.GetAwaiter().GetResult(); // rethrow exceptions
        }
        else
        {
            write.AsTask().GetAwaiter().GetResult();
        }
    }
}
catch (Exception ex)
{
}
```

Reply:
If structuring like this, you'd want to take advantage of `.IsCompletedSuccessfully` and use the `.Result` directly more cheaply, but note: this is the exception case in general where a write exceeded the buffer.

Reply:
There is no `.Result` on `ValueTask` (only on `ValueTask<T>`). But that is fine - `GetAwait().GetResult()` is 0 allocation. So this call here is only to rethrow exceptions, if any. One exception is the writer has completed before accepting your item.

Reply:
I think this might write the log entry twice? Because no cancellation token is passed to WriteAsync, so its still going to try to write even after we gave up waiting. I think to address that I'd need to stop using WaitAsync, and instead create a cancellation token source and use cancel after. I?

Reply:
Assuming we want to avoid duplicate writes, what is the downside of this approach?

```csharp
            try
            {
                if (_consoleBuffer.Writer.TryWrite(evt) == false)
                {
                    // Buffer is currently full, wait until writing is permitted.
                    var source = new CancellationTokenSource(_consoleBufferTimeout);
                    var writeTask = _consoleBuffer.Writer.WriteAsync(evt, source.Token);

                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task if we want to wait for the buffer to clear.
                    writeTask.GetAwaiter().GetResult();
                }
            }
            catch (Exception ex)
            {
                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
                // dump the raw exception and write the event to the console directly
                _exceptionhandler(ex);
                Console.WriteLine(evt);
            }
```